suppressPackageStartupMessages(library(pacman))
suppressPackageStartupMessages(p_load(tidyverse))
use.reticulate <- FALSE
if(use.reticulate) {
    p_load(reticulate)
} else {
    p_load(synapser)    
}
p_load(pheatmap)
suppressPackageStartupMessages(p_load(reshape2))
suppressPackageStartupMessages(p_load(plyr))
suppressPackageStartupMessages(p_load(dplyr))


# `synasper` not working on local computer, so using `reticulate` to
# log in to Synapse instead
if(use.reticulate) {
    use_condaenv("synapse-r")
    synapseclient <- reticulate::import('synapseclient')
    syn <- synapseclient$Synapse()
    syn$login(silent = T)
} else {
    synLogin()
}

suppressPackageStartupMessages(p_load("foreach"))
suppressPackageStartupMessages(p_load("parallel"))
suppressPackageStartupMessages(p_load("ComplexHeatmap"))

num.cores <- detectCores()
if(!is.na(num.cores) && (num.cores > 1)) {
  suppressPackageStartupMessages(p_load("doMC"))
  cat(paste("Registering ", num.cores-1, " cores.\n", sep=""))
  registerDoMC(cores=(num.cores-1))
}
num.processes <- num.cores - 1


phases <- c("training", "leaderboard", "validation")
names(phases) <- phases

limit.genes.to.protein.coding <- function(genes, use.symbols = TRUE) {

    use.biomaRt <- TRUE
    exclude.mt <- FALSE
    if(use.biomaRt) {
        suppressPackageStartupMessages(p_load(biomaRt))
        ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
        pc.tbl <- getBM(attributes = c("ensembl_gene_id","external_gene_name","description","chromosome_name"),
                        filters = 'biotype', values = c('protein_coding'), mart = ensembl)
        if(exclude.mt) {
            pc.tbl <- subset(pc.tbl, chrosome_name != "MT")
        }
        keys <- as.character(pc.tbl$external_gene_name)
        if(!use.symbols) { keys <- as.character(pc.tbl$ensembl_gene_id) }
        genes <- genes[genes %in% keys]
        return(genes)
    }
    if(exclude.mt) { stop("Have not implemented exclude.mt for annotationHub -- but is easy\n") }
    suppressPackageStartupMessages(p_load(AnnotationHub))
    suppressPackageStartupMessages(p_load(ensembldb))
    ah <- AnnotationHub()
    flag <- (ah$species == "Homo sapiens") & (ah$genome == "GRCh38") & (ah$dataprovider == "Ensembl") & (ah$rdataclass == "EnsDb")
    ah2 <- ah[flag, ]
    ## as.data.frame(mcols(ah2))[1:10,c("title"),drop=FALSE]
    edb <- ah2[["AH73881"]]

    ## keytypes(edb)
    ## columns(edb)
    keys <- keys(edb, "GENENAME")
    columns <- c("GENEID", "ENTREZID", "GENEBIOTYPE")
    tbl <- ensembldb::select(edb, keys, columns, keytype = "GENENAME")
    pc.tbl <- subset(tbl, GENEBIOTYPE == "protein_coding")

    keys <- as.character(pc.tbl$GENENAME)
    if(!use.symbols) { keys <- as.character(pc.tbl$GENEID) }
    genes <- genes[genes %in% keys]
    return(genes)
}

read.csv.from.synapse <- function(synId) {
    if(use.reticulate) {
        file <- syn$get(synId)$path
    } else {
        file <- synGet(synId)$path
    }
    ret <- suppressMessages(read_csv(file))
}

read.csv.table.from.synapse <- function(synId, sep=",") {
    if(use.reticulate) {
        file <- syn$get(synId)$path
    } else {
        file <- synGet(synId)$path
    }
    ret <- suppressMessages(read.table(file, sep=sep, header=TRUE))
}


expr.synIds <- list("training" = "syn21212911",
                    "leaderboard" = "syn21671825",
                    "validation" = "syn21212718")

## Read in the expression matrices
raw.exprs <-
    llply(expr.synIds,
          .fun = function(synId) {
              # RNAseq data
              if(use.reticulate) {
                  rnaseq_file <- syn$get(synId)$path
              } else {
                  rnaseq_file <- synGet(synId)$path
              }
              rnaseq <- suppressMessages(read_csv(rnaseq_file))
              rnaseq
          })

## Read in the intermediate results generated by sc1-analysis-preprocessing.R

gene.symbol.expr.cor.synIds <-
    list("training" = "syn22397862",
         "leaderboard" = "syn22398045",
         "validation" = "syn22398135")

gene.symbol.expr.cors <-
    llply(gene.symbol.expr.cor.synIds,
          .fun = function(synId) read.csv.from.synapse(synId))

symbol.expr.synIds <-
    list("training" = "syn22397592",
         "leaderboard" = "syn22397931",
         "validation" = "syn22398088")

symbol.exprs <-
    llply(symbol.expr.synIds,
          .fun = function(synId) read.csv.table.from.synapse(synId))

auc.synIds <-
    list("training" = "syn22397379",
         "leaderboard" = "syn22397867",
         "validation" = "syn22398053")

aucs <-
    llply(auc.synIds,
          .fun = function(synId) read.csv.table.from.synapse(synId))

## Read in the highly variable and expressed genes from our FIMM/OHSU Beat AML manuscript
synId <- "syn21899141"
synId <- "syn20628997"
aml.expressed.variable.genes.tbl <- read.csv.table.from.synapse(synId, sep="\t")
gene_symbols <- as.data.frame(unique(raw.exprs[["training"]][, c("Gene", "Symbol")]))
aml.expressed.variable.symbols <- subset(gene_symbols, Gene %in% as.character(aml.expressed.variable.genes.tbl$gene))$Symbol

filter.genes <- limit.genes.to.protein.coding(aml.expressed.variable.symbols)

## Read in the drug family annotations
drug.families <- read.table("41586_2018_623_MOESM3_ESM-s11.txt", sep="\t", header=TRUE, stringsAsFactors=FALSE)
drug.families$value <- "Yes"
drug.family.mat <- acast(data = drug.families, formula = family ~ inhibitor, fill = NA)
drug.family.mat <- as.data.frame(drug.family.mat)

## Define expressed genes from the distribution of the mean expression
## of the _training_ data
mu <- apply(symbol.exprs[["training"]], 1, mean)
SD <- apply(symbol.exprs[["training"]], 1, sd)

cutoff <- 0

plot(density(mu))
abline(v=cutoff)

## filter.genes <- names(mu)[mu > cutoff]

## Read in the prediction results from the validation phase
synId <- "syn22267537"
validation.result.tbl <- synTableQuery(paste0("SELECT * FROM ", synId))
validation.result.tbl <- as.data.frame(validation.result.tbl)

## Exclude "gold" (which must mean gold-standard because its pearson and spearman = 1)
validation.result.tbl <- subset(validation.result.tbl, team != "gold")
validation.result.tbl$pearson <- as.numeric(validation.result.tbl$pearson)

drug.mean.validation.result.tbl <-
    ddply(validation.result.tbl, .variables = c("inhibitor"),
          .fun = function(df) data.frame(pearson = mean(df$pearson, na.rm=TRUE)))

save.image(".Rdata")

gene.symbol.expr.cors.bh <-
    llply(gene.symbol.expr.cors,
          .fun = function(cors) {
              ddply(subset(cors, gene %in% filter.genes), .variables = c("drug"),
                    .fun = function(df) {
                        data.frame(gene = df$gene, cor=df$pearson.cor,
                                   pval = df$pearson.pval,
                                   padj = p.adjust(df$pearson.pval))
                    })
          })

gene.symbol.expr.cors.sig <-
    llply(gene.symbol.expr.cors.bh,
          .fun = function(df) subset(df, padj < 0.05))

gene.symbol.expr.cors.sig.mats <-
    llply(gene.symbol.expr.cors.sig,
          .fun = function(sig.cors) {
              sig.cor.mat <-
                  acast(data = sig.cors[, c("drug", "gene", "cor")],
                        formula = gene ~ drug)
          })

sig.cor.mat <- gene.symbol.expr.cors.sig.mats[["training"]]

tmp.mat <- gene.symbol.expr.cors.sig[["training"]]
drugs.to.exclude <- c("Venetoclax")
tmp.mat <- subset(tmp.mat, !(drug %in% drugs.to.exclude))
sig.cor.genes <- as.character(unique(tmp.mat$gene))
sig.cors.all <- subset(gene.symbol.expr.cors[["training"]], gene %in% sig.cor.genes)
sig.cor.mat.all <- acast(data = sig.cors.all[, c("drug", "gene", "pearson.cor")],
                         formula = gene ~ drug)

l <- list(drug.family.mat, sig.cor.mat, sig.cor.mat.all)
all.drugs <- unique(unlist(lapply(l, colnames)))
all.drugs <- unique(c(all.drugs, gene.symbol.expr.cors$training$drug))

rename.drug.map <- data.frame(drug = unique(all.drugs),
                              safe.name = make.names(unique(all.drugs)),
                              stringsAsFactors = FALSE)

rownames(rename.drug.map) <- rename.drug.map$safe.name

renamed.aucs <- aucs
for(nm in names(renamed.aucs)) {
    missing <- colnames(renamed.aucs[[nm]])[!(colnames(renamed.aucs[[nm]]) %in% rownames(rename.drug.map))]
    if(length(missing) > 0) {
        cat(paste0("Missing some drugs: ", paste(missing, sep=", "), "\n"))
    }
    renamed.aucs[[nm]] <- renamed.aucs[[nm]][, colnames(renamed.aucs[[nm]]) %in% rownames(rename.drug.map)]
    colnames(renamed.aucs[[nm]]) <- rename.drug.map[colnames(renamed.aucs[[nm]]), "drug"]
}

auc.ranges <-
    llply(renamed.aucs,
          .fun = function(aucs) {
              apply(aucs, 2, mad, na.rm=TRUE)
          })

l <- list(drug.family.mat, sig.cor.mat, sig.cor.mat.all, renamed.aucs[["training"]])
all.drugs <- unique(unlist(lapply(l, colnames)))

rownames(drug.mean.validation.result.tbl) <- drug.mean.validation.result.tbl$inhibitor
missing.predictions <- all.drugs[!(all.drugs %in% rownames(drug.mean.validation.result.tbl))]

prediction.drugs <- rownames(drug.mean.validation.result.tbl)
all.drugs <- intersect(all.drugs, prediction.drugs)
mean.drug.predictions <- drug.mean.validation.result.tbl$pearson
names(mean.drug.predictions) <- rownames(drug.mean.validation.result.tbl)

all.drugs <- unique(c(all.drugs, "GRD"))
all.drugs <- all.drugs[!(all.drugs %in% drugs.to.exclude)]

drug.cor.mat <- cor(renamed.aucs[["training"]], use = "pairwise.complete.obs")

sig.cor.mat <- as.data.frame(sig.cor.mat)
sig.cor.mat.all <- as.data.frame(sig.cor.mat.all)
drug.cor.mat <- as.data.frame(drug.cor.mat)

training.auc.ranges <- auc.ranges[["training"]]



for(d in all.drugs) {
    if(!(d %in% colnames(drug.family.mat))) { drug.family.mat[, d] <- NA }
    if(!(d %in% colnames(sig.cor.mat))) { sig.cor.mat[, d] <- NA }
    if(!(d %in% colnames(drug.cor.mat))) { drug.cor.mat[, d] <- NA }    
    if(!(d %in% colnames(sig.cor.mat.all))) { sig.cor.mat.all[, d] <- NA }
    if(!(d %in% names(mean.drug.predictions))) { mean.drug.predictions[d] <- NA }
    if(!(d %in% names(training.auc.ranges))) { training.auc.ranges[d] <- NA }    
    
}    
drug.family.mat <- drug.family.mat[, all.drugs]
sig.cor.mat <- sig.cor.mat[, all.drugs]
drug.cor.mat <- drug.cor.mat[all.drugs, all.drugs]
sig.cor.mat.all <- sig.cor.mat.all[, all.drugs]
mean.drug.predictions <- mean.drug.predictions[all.drugs]
training.auc.ranges <- training.auc.ranges[all.drugs]
num.sig.genes <- colSums(!is.na(sig.cor.mat))

make.plot.dendro <- function(mat, sig.vec, perf.vec, range.vec, cor.mat, drug.gene.cor.mat, column_dend, row_dend,
                             column_split, row_split, ...) {
    breaks = c(1, 10, 100, 1000)
    breaks <- breaks[breaks < max(sig.vec)]
##    if(max(sig.vec) > max(breaks)) { breaks <- c(breaks, max(sig.vec)) }
    log.breaks <- c(0, log10(breaks))
    label.breaks <- c(0, breaks)
    vals <- log10(as.numeric(sig.vec))
    vals[is.infinite(vals)] <- 0
    n.targets <- colSums(!is.na(mat))    
    col_ha = HeatmapAnnotation(perf = anno_barplot(perf.vec),
                               range = anno_barplot(range.vec),
                               sig = anno_barplot(vals,
                                                  axis_param = list(at = log.breaks, labels = label.breaks)),
                               targets = anno_barplot(n.targets))
                               
    names(col_ha) = c("Mean Pearson", "Drug Range (MAD)", "# Sig Genes", "# Targets")
    mat <- as.matrix(mat)
    cor.mat <- as.matrix(cor.mat)
    drug.gene.cor.mat <- as.matrix(drug.gene.cor.mat)
    row.flag <- rowSums(is.na(mat)) != ncol(mat)
    colors <- structure(1, names = c("Yes"))
    h <- Heatmap(mat[row.flag,], cluster_rows = FALSE, cluster_columns = column_dend, top_annotation = col_ha,
                 column_names_rot = 45, na_col = "white", show_heatmap_legend = FALSE, 
                 row_names_gp = gpar(fontsize = 4),
                 column_names_gp = gpar(fontsize = 4),
                 show_column_dend = TRUE, column_split = column_split,
                 height = 1, col = colors,
                 ...)
    h2 <- Heatmap(cor.mat, cluster_rows = row_dend, cluster_columns = column_dend, show_row_dend = FALSE,
                  name = "Drug/drug\ncorrelation",
                  column_split = column_split, row_split = row_split,
                  row_names_gp = gpar(fontsize = 4),
                  show_row_names = FALSE,
                  row_title = "Drug",
                  height = 1,
                  ...)
    h3 <- Heatmap(drug.gene.cor.mat, cluster_rows = TRUE, cluster_columns = column_dend,
                  show_row_dend = FALSE,
                  name = "Drug/gene\ncorrelation",
                  column_split = column_split,
                  column_names_rot = 45,
                  column_names_gp = gpar(fontsize = 4),
                  row_names_gp = gpar(fontsize = 4),
                  show_row_names = FALSE,
                  row_title = "Gene",
                  column_title = "Drug",
                  column_title_side = "bottom",
                  height = 1,
                  ...)
    ht_list = h %v% h2 %v% h3
    draw(ht_list, column_title = "Drug", column_title_side = "bottom")
}

drugs <- all.drugs

drug.mat <- drug.cor.mat[drugs, drugs]

column.hc <- hclust(dist(drug.mat, method="euclidean"), method="complete")
column_dend <- as.dendrogram(column.hc)
row_dend <-
    as.dendrogram(hclust(dist(t(drug.mat), method="euclidean"),
                         method="complete"))

num.clusters <- 7
ct <- cutree(column.hc, k=num.clusters)
drug.clusters <- data.frame(inhibitor = names(ct), cluster = unname(ct))

png("sc1-training-data-vs-validation-performance.png")
make.plot.dendro(drug.family.mat[, drugs],
                 num.sig.genes[drugs],
                 mean.drug.predictions[drugs],
                 training.auc.ranges[drugs],
                 drug.cor.mat[drugs, drugs],
                 sig.cor.mat.all[, drugs],
                 column_dend,
                 row_dend,
                 column_split = num.clusters,
                 row_split = num.clusters,                 
                 column_title_gp = gpar(fill = c("red", "blue", "green", "orange", "purple", "yellow", "brown"), font = 1:7))
d <- dev.off()

drug.range.tbl <- data.frame(inhibitor = names(training.auc.ranges), range = unname(training.auc.ranges))
mean.perf.tbl <- data.frame(inhibitor = names(mean.drug.predictions), mean.correlation = unname(mean.drug.predictions))

all.tbl <- merge(drug.clusters, drug.range.tbl, all.x = TRUE)
all.tbl <- merge(all.tbl, mean.perf.tbl, all.x = TRUE)
all.tbl <- all.tbl[!(all.tbl$inhibitor == "GRD"),]


suppressPackageStartupMessages(p_load(ggplot2))
suppressPackageStartupMessages(p_load(gridExtra))
all.tbl$cluster <- factor(all.tbl$cluster, levels = sort(unique(all.tbl$cluster)))
g1 <- ggplot(data = all.tbl)
g1 <- g1 + geom_boxplot(aes(x = cluster, y = mean.correlation))
g1 <- g1 + ylab("Mean Pearson Correlation\n(over Teams)")

g2 <- ggplot(data = all.tbl)
g2 <- g2 + geom_boxplot(aes(x = cluster, y = range))
g2 <- g2 + ylab("Dynamic Range\n(AUC MAD over Samples)")

png("sc1-training-clusters-vs-validation-performance.png")
grid.arrange(g1,g2)
d <- dev.off()
